contributors/design-proposals/sig-cli/decoupling_kubectl_from_k8s.md

```diff
@@ -0,0 +1,244 @@
+# Decoupling kubectl from core k8s
+
+Status: Pending
+
+Version: Alpha
+
+Implementation Owner: jregan@google.com
+
+## Motivation
+
+The [kubectl program] is a command-line client used to configure and control k8s
+(kubernetes) clusters.  The program lives in the core [k8s repo].
+
+kubectl depends on various parts of k8s and k8s depends on various parts of
+kubectl with varying degrees of intentional design.  If a bug is detected in how
+kubectl interacts with k8s, there's no procedure to release a fix that doesn't
+involve a full release of k8s.
+
+This document describes how a clear line will be drawn in the code between
+kubectl and the remainder of k8s to leave only intentional dependencies.
+
+This is a precondition to being able to release kubectl as a product distinct
+from the rest of k8s, or more dramatic changes
+like
+[moving kubectl out of the kubernetes repository][proposal for kubectl repo] and
+into [its own repository][kubectl repo].
+
+## Proposal
+
+1. Identify, via `BUILD` file rules, existing good and bad dependencies.
+
+   _Good dependencies are on public libraries intended for the given usage; bad
+   dependencies are otherwise._
+
+   At the same time, inhibit introduction of _new_ bad dependencies.
+
+1. Create. delete, move or merge packages to _remove_ bad dependencies.
+
+   At the same time, adapt CICD tools, miscellaneous shell scripts, etc. to
+   continue to function properly.
+
+
+![insert amazing before and after diagram here][reposCandide]
+
+### Non-goals
+
+#### Extensive changes to testing.
+
+Decoupling kubectl is a necessary but insufficient condition for releasing
+kubectl at a different cadence from k8s.
+
+Another necessary condition is an end to end (e2e) test framework to test an
+instance of kubectl (previously released or HEAD) against an instance of k8s
+(likewise, previously released or HEAD).  This may require its own proposal if
+existing framework cannot handle it.
+
+#### Immediately changing policy to release kubectl and k8s at a different cadence.
+
+This can wait.  For purposes of bootstrapping a new style of e2e testing and
+release, its convenient to already have a _decoupled_ kubectl in the wild,
+released per the _existing_ process.  Any new process will need a rollback state
+it understands.
+
+#### Moving kubectl to its [own repo][kubectl repo], outside the [k8s repo].
+
+This decoupling work is a precondition to such a move, but doesn't _require_ a
+new repo.
+
+## User Experience
+
+The user should see no change other than, it's hoped, faster feature development
```

Comment:
The decoupling may have some limited UX changes - e.g. switching to versioned clients from unversioned clients